### PR TITLE
MTC-2953 Add the option to set the content of tokens in email_stats to NULL and the improvement that only email_stats entries will be cleaned that referenced emails entry is currently not published

### PR DESCRIPTION
**Pull request by beetofly**
## Description
This feature adds 

- the option to the leuchtfeuer housekeeping bundle to set the content of tokens fields in email_stats to NULL to "reduce the weight" of the entries and

- the improvement that only email_stats entries will be cleaned that referenced emails entry is currently not published.

Important notice: If a referenced email is ever switched back to published, the contacts will get the email again. 

## How to test
Apply the patch of this PR in a mautic testenvironment with the leuchtfeuer housekeeping bundle. 

1.) Run for example `php bin/console mautic:leuchtfeuer:housekeeping -t -d 50`

2.) After that check the table email_stats -> the content in `tokens` for all rows where date_sent is older than 50 days have been set to NULL.

3.) Run for example `php bin/console mautic:leuchtfeuer:housekeeping -m -d 50`

4.) After that check the table email_stats -> all entries are gone that referenced email_id is NULL or that referenced email is not published and that are older than 50 days.

5.) If you want to do a dry run for all that, add the option flag -r.

## Notice:
This option `--email-stats-tokens` (shortcut -t) can not be combined with any "-c", "-l" or "-m" flag in one command.
If you do so, it will throw you an error with the notice.

### CODE REVIEW

File: Command/EventLogCleanupCommand.php [hunk 1]
```diff
@@ -42,13 +42,14 @@ protected function configure(): void
                     new InputOption('dry-run', 'r', InputOption::VALUE_NONE, 'Do a dry run without actually deleting anything.'),
                     new InputOption('campaign-lead', 'c', InputOption::VALUE_NONE, 'Purge only Campaign Lead Event Log Records'),
                     new InputOption('lead', 'l', InputOption::VALUE_NONE, 'Purge only Lead Event Log Records'),
-                    new InputOption('email-stats', 'm', InputOption::VALUE_NONE, 'Purge only Email Stats Records + Email Stats Devices'),
+                    new InputOption('email-stats', 'm', InputOption::VALUE_NONE, 'Purge only Email Stats Records the referenced emails entry is currently not published and purge Email Stats Devices. Important: If referenced email is ever switched back to published, the contacts will get the email again.'),
```
Comment (Moongazer) [inline Command/EventLogCleanupCommand.php:45]: Correction: "Purge only Email Stats Records **where** the referenced email~~s~~ entry..."

File: Command/EventLogCleanupCommand.php [hunk 2]
```diff
@@ -42,13 +42,14 @@ protected function configure(): void
...
+                    new InputOption('email-stats-tokens', 't', InputOption::VALUE_NONE, 'Set only tokens fields in Email Stats Records to NULL. Important: This option can not be combined with any "-c", "-l" or "-m" flag in one command. And: If the option flag "-t" is not set, the NULL setting of tokens will not be done with the basis command, so if you just run mautic:leuchtfeuer:housekeeping without a flag)'),
                     new InputOption('cmp-id', 'i', InputOption::VALUE_OPTIONAL, 'Delete only campaign_lead_eventLog for a specific CampaignID', 'none'),
                 ]
             )
             ->setHelp(
                 <<<'EOT'
-                The <info>%command.name%</info> command is used to clean up the campaign_lead_event_log, lead_event_log, email_stats and email_stats_devices table.
+                The <info>%command.name%</info> command is used to clean up the campaign_lead_event_log table, the lead_event_log table, the email_stats table (but only email_stats entries the referenced emails entry is currently not published) and the email_stats_devices table or just clean up the field tokens in email_stats if the option flag "-t" is set.
```
Comment (Moongazer) [inline Command/EventLogCleanupCommand.php:52]: Correction: "(but only email_stats entries **where** the referenced email~~s~~ entry..."

File: Command/EventLogCleanupCommand.php [hunk 3]
```diff
@@ -67,8 +68,11 @@ protected function configure(): void
                 Purge only lead_event_log records
                 <info>php %command.full_name% --lead</info> 
                 
-                Purge only email_stats and email_stats_devices records:
-                <info>php %command.full_name% --email-stats </info> 
+                Purge only email_stats the referenced emails entry is currently not published and email_stats_devices records [Important: If referenced email is ever switched back to published, the contacts will get the email again]:
```
Comment (Moongazer) [inline Command/EventLogCleanupCommand.php:71]: Correction: "Purge only email_stats **where** the referenced email~~s~~ entry..."

File: Command/EventLogCleanupCommand.php [hunk 4]
```diff
@@ -82,10 +86,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+            EventLogCleanup::EMAIL_STATS_TOKENS   => $input->getOption('email-stats-tokens')
         ];
 
         if (0 === array_sum($operations)) {
             $operations = array_combine(array_keys($operations), array_fill(0, count($operations), true));
+            array_pop($operations);
+            $operations["email_stats_tokens"] = false;
+        }
+
+        if (($operations["email_stats_tokens"] === true) && ((($operations["email_stats"] === true) || ($operations["campaign_lead_event_log"] === true)) || ($operations["lead_event_log"] === true))) {
```
Comment (Moongazer) [inline Command/EventLogCleanupCommand.php:98]: Same here: use the constants as array keys

File: README.md [hunk 5]
```diff
@@ -3,25 +3,28 @@ This bundle provides a Mautic Housekeeping Command for database cleanup purposes
 
 ## Command
 
-Command to delete lead_event_log, campaign_lead_event_log, email_stats and email_stats_devices table entries. 
+Command to delete lead_event_log table entries, campaign_lead_event_log table entries, email_stats table entries that referenced emails entry is currently not published and email_stats_devices table entries.
```
Comment (Moongazer) [inline README.md:6]: Correction: "Command to delete lead_event_log, campaign_lead_event_log, email_stats_devices and email_stats table entries where the referenced email entry is currently not published."

File: README.md [hunk 6]
```diff
@@ -3,25 +3,28 @@ This bundle provides a Mautic Housekeeping Command for database cleanup purposes
...
+
+Important: If referenced email is ever switched back to published, the contacts will get the email again.
 
 ```
 bin/console mautic:leuchtfeuer:housekeeping
 ```
-By default, entries older than 365 days are deleted from the CampaignLeadEventLog, LeadEventLog, EmailStats and EmailStatsDevices tables.
+By default, entries older than 365 days are deleted from the CampaignLeadEventLog, LeadEventLog, EmailStats (only email_stats entries that referenced emails entry is currently not published) and EmailStatsDevices tables.
```
Comment (Moongazer) [inline README.md:13]: Correction: "(only email_stats entries **where** the referenced email~~s~~ entry is currently not published)"

File: Service/EventLogCleanup.php [hunk 7]
```diff
@@ -92,10 +103,16 @@ public function deleteEventLogEntries(int $daysOld, ?int $campaignId, bool $dryR
             $operations[self::EMAIL_STATS]         = true;
         }
 
+        if (array_key_exists(self::EMAIL_STATS_TOKENS, $operations) && true === $operations[self::EMAIL_STATS_TOKENS]) {
+            unset($operations[self::EMAIL_STATS_TOKENS]);
+            $operations[self::EMAIL_STATS_TOKENS]  = true;
+        }
+
```
Comment (Moongazer) [inline Service/EventLogCleanup.php:110]: Not sure about what's happening here, please explain sense:

```
if:
  EMAIL_STATS_TOKENS key exists in $operations array AND array value of EMAIL_STATS_TOKENS key is TRUE
then:
  unset value of array key EMAIL_STATS_TOKENS;
  set value of array key EMAIL_STATS_TOKENS to TRUE again;
```

File: Service/EventLogCleanup.php [hunk 8]
```diff
@@ -107,9 +124,16 @@ public function deleteEventLogEntries(int $daysOld, ?int $campaignId, bool $dryR
                     continue;
                 }
 
-                $sql                     = 'SELECT * FROM '.str_replace(self::PREFIX, $this->dbPrefix, $this->queries[$operation]);
-                $statement               = $this->connection->executeQuery($sql, $this->params[$operation], $this->types[$operation]);
-                $result[$operation]      = $statement->rowCount();
+                if (true === $operations['email_stats_tokens']) {
```
Comment (Moongazer) [inline Service/EventLogCleanup.php:127]: Use constant key `self::EMAIL_STATS_TOKENS`

File: Service/EventLogCleanup.php [hunk 9]
```diff
@@ -121,7 +145,13 @@ public function deleteEventLogEntries(int $daysOld, ?int $campaignId, bool $dryR
                     continue;
                 }
 
-                $sql                     = 'DELETE '.$this->dbPrefix.$operation.' FROM '.str_replace(self::PREFIX, $this->dbPrefix, $this->queries[$operation]);
+                if (true === $operations['email_stats_tokens']) {
```
Comment (Moongazer) [inline Service/EventLogCleanup.php:148]: Use constant key `self::EMAIL_STATS_TOKENS`

File: Service/EventLogCleanup.php [hunk 10]
```diff
@@ -155,9 +185,11 @@ public function deleteEventLogEntries(int $daysOld, ?int $campaignId, bool $dryR
 
             $message .= $result[$operation].' '.$operation;
         }
-
-        $message .= $dryRun ? $this->dryRunMessage : $this->runMessage;
-
+        if (true === $operations['email_stats_tokens']) {
```
Comment (Moongazer) [inline Service/EventLogCleanup.php:188]: Use constant key `self::EMAIL_STATS_TOKENS`

File: Command/EventLogCleanupCommand.php [hunk 11]
```diff
@@ -82,10 +86,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             EventLogCleanup::CAMPAIGN_LEAD_EVENTS => $input->getOption('campaign-lead'),
             EventLogCleanup::LEAD_EVENTS          => $input->getOption('lead'),
             EventLogCleanup::EMAIL_STATS          => $input->getOption('email-stats'),
+            EventLogCleanup::EMAIL_STATS_TOKENS   => $input->getOption('email-stats-tokens')
         ];
 
         if (0 === array_sum($operations)) {
             $operations = array_combine(array_keys($operations), array_fill(0, count($operations), true));
+            array_pop($operations);
```
Comment (Moongazer) [inline Command/EventLogCleanupCommand.php:94]: Remove this line

**Review by Moongazer — CHANGES_REQUESTED** [review]
(See comments in code lines)

File: Command/EventLogCleanupCommand.php (hunk 1 above)
Comment (beetofly) [reply Command/EventLogCleanupCommand.php]: Thanks @moongazer for this spelling recommendation - implemented that with my current commit.

File: Command/EventLogCleanupCommand.php (hunk 2 above)
Comment (beetofly) [reply Command/EventLogCleanupCommand.php]: Thanks @moongazer for this spelling recommendation - implemented that with my current commit.

File: Command/EventLogCleanupCommand.php (hunk 3 above)
Comment (beetofly) [reply Command/EventLogCleanupCommand.php]: Thanks @moongazer for this spelling recommendation - implemented that with my current commit.

File: Command/EventLogCleanupCommand.php (hunk 11 above)
Comment (beetofly) [reply Command/EventLogCleanupCommand.php]: Thanks @moongazer for the recommendation, that  `array_pop($operations)` is not needed - removed this not harmful but not necessary  part with my current commit.

File: Command/EventLogCleanupCommand.php [hunk 12]
```diff
@@ -82,10 +86,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+            $operations["email_stats_tokens"] = false;
```
[the remark this answers is not in the document]
Comment (beetofly) [reply Command/EventLogCleanupCommand.php]: Thanks @moongazer for the recommendation to better use constans if we have them - changed this not harmful part in the better practise part with my current commit.

File: Command/EventLogCleanupCommand.php (hunk 4 above)
Comment (beetofly) [reply Command/EventLogCleanupCommand.php]: Thanks @moongazer for the recommendation to better use constans if we have them - changed this not harmful part in the better practise part with my current commit.

File: README.md (hunk 5 above)
Comment (beetofly) [reply README.md]: Thanks @Moongazer for this spelling recommendation - implemented that with my current commit.

File: README.md (hunk 6 above)
Comment (beetofly) [reply README.md:13]: Thanks @Moongazer for this spelling recommendation - implemented that with my current commit.

File: Service/EventLogCleanup.php (hunk 7 above)
Comment (beetofly) [reply Service/EventLogCleanup.php]: Talked with @Moongazer about this.

Result: Thanks @Moongazer for the recommendation, that this part is not necessarrely needed (didn´t realized it in the first step) - so I removed this not harmful but not necessary part with my current commit.

File: Service/EventLogCleanup.php (hunk 8 above)
Comment (beetofly) [reply Service/EventLogCleanup.php]: Thanks @Moongazer for the recommendation to better use constans if we have them - changed this not harmful part in the better practise part with my current commit.

File: Service/EventLogCleanup.php (hunk 9 above)
Comment (beetofly) [reply Service/EventLogCleanup.php]: Thanks @Moongazer for the recommendation to better use constans if we have them - changed this not harmful part in the better practise part with my current commit.

File: Service/EventLogCleanup.php (hunk 10 above)
Comment (beetofly) [reply Service/EventLogCleanup.php]: Thanks @Moongazer for the recommendation to better use constans if we have them - changed this not harmful part in the better practise part with my current commit.

**Review by Moongazer — APPROVED** [review]

**Review by oltmanns-leuchtfeuer — APPROVED** [review]
LGTM!

**Pull request merged**